### PR TITLE
respect the record format when copying a store to edge machines

### PR DESCRIPTION
**Pull request by mneedham**

### CODE REVIEW

**Comment by davidegrohmann** [comment]
@mneedham can we add a test ensuring that we fail if the copied store has the wrong format?

**Comment by mneedham** [comment]
@davidegrohmann  `ConvertNonCoreEdgeStoreIT` fails if you have the wrong format but I can put something directly against `CopiedStoreRecovery` as well

**Comment by davidegrohmann** [comment]
@mneedham Not sure if we need another test or not. The scenario I have in mind is that a new core with a misconfigured store format enters the cluster and it should fail when it gets the store (through store copy) from another core in the cluster.

**Comment by mneedham** [comment]
Hmmm that test fails. Will try locally

**Comment by mneedham** [comment]
@davidegrohmann what you described is pretty much the same as the problem we were having with the edge machines in the soak test not starting up. I can add a test that specifically tests for that exception though. 

It'd be good to make that failure case explicit.

**Comment by mneedham** [comment]
@davidegrohmann I've added an IT that tests for this and then tweaked the copying code to put a bit clearer message if it fails.

**Comment by mneedham** [comment]
@davidegrohmann weird, it's failing on the ITs when trying to do something with Lucene. Works locally though...https://build.neohq.net/viewLog.html?buildId=712776&buildTypeId=PullRequests_PullRequests

**Comment by davidegrohmann** [comment]
@mneedham the error is about file descriptors from the logs:

```
org.neo4j.kernel.lifecycle.LifecycleException: Component 'org.neo4j.kernel.impl.storageengine.impl.recordstorage.RecordStorageEngine@1f7ef57d' was successfully initialized, but failed to start. Please see attached cause exception.
    at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:443)
    at org.neo4j.kernel.lifecycle.LifeSupport.start(LifeSupport.java:107)
    at org.neo4j.kernel.NeoStoreDataSource.start(NeoStoreDataSource.java:506)
    at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:433)
    at org.neo4j.kernel.lifecycle.LifeSupport.start(LifeSupport.java:107)
    at org.neo4j.kernel.impl.transaction.state.DataSourceManager.start(DataSourceManager.java:99)
    at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:433)
    at org.neo4j.kernel.lifecycle.LifeSupport.start(LifeSupport.java:107)
    at org.neo4j.kernel.impl.factory.GraphDatabaseFacadeFactory.initFacade(GraphDatabaseFacadeFactory.java:140)
    at org.neo4j.kernel.impl.enterprise.EnterpriseFacadeFactory.initFacade(EnterpriseFacadeFactory.java:42)
    at org.neo4j.kernel.impl.factory.GraphDatabaseFacadeFactory.newFacade(GraphDatabaseFacadeFactory.java:108)
    at org.neo4j.com.storecopy.ExternallyManagedPageCache$GraphDatabaseFactoryWithPageCacheFactory.newDatabase(ExternallyManagedPageCache.java:129)
    at org.neo4j.graphdb.factory.GraphDatabaseFactory.lambda$createDatabaseCreator$192(GraphDatabaseFactory.java:88)
    at org.neo4j.graphdb.factory.GraphDatabaseFactory$$Lambda$237/1820383114.newDatabase(Unknown Source)
    at org.neo4j.graphdb.factory.GraphDatabaseBuilder.newGraphDatabase(GraphDatabaseBuilder.java:183)
    at org.neo4j.coreedge.catchup.storecopy.edge.CopiedStoreRecovery.newTempDatabase(CopiedStoreRecovery.java:91)
    at org.neo4j.coreedge.catchup.storecopy.edge.CopiedStoreRecovery.recoverCopiedStore(CopiedStoreRecovery.java:52)
    at org.neo4j.coreedge.catchup.storecopy.LocalDatabase.copyStoreFrom(LocalDatabase.java:92)
    at org.neo4j.coreedge.server.edge.EdgeServerStartupProcess.start(EdgeServerStartupProcess.java:87)
    at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:433)
    at org.neo4j.kernel.lifecycle.LifeSupport.start(LifeSupport.java:107)
    at org.neo4j.kernel.impl.factory.GraphDatabaseFacadeFactory.initFacade(GraphDatabaseFacadeFactory.java:140)
    at org.neo4j.coreedge.server.EnterpriseCoreEdgeFacadeFactory.initFacade(EnterpriseCoreEdgeFacadeFactory.java:38)
    at org.neo4j.coreedge.server.EnterpriseEdgeFacadeFactory.initFacade(EnterpriseEdgeFacadeFactory.java:30)
    at org.neo4j.coreedge.server.edge.EdgeGraphDatabase.<init>(EdgeGraphDatabase.java:41)
    at org.neo4j.coreedge.discovery.Cluster.startEdgeServer(Cluster.java:306)
    at org.neo4j.coreedge.discovery.Cluster.startEdgeServer(Cluster.java:282)
    at org.neo4j.coreedge.discovery.Cluster.addEdgeServerWithFileLocation(Cluster.java:436)
    at org.neo4j.coreedge.scenarios.ConvertNonCoreEdgeStoreIT.shouldReplicateTransactionToCoreServers(ConvertNonCoreEdgeStoreIT.java:117)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:497)
    at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
    at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
    at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
    at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
    at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
    at org.neo4j.test.rule.TargetDirectory$TestDirectory$1.evaluate(TargetDirectory.java:133)
    at org.junit.rules.RunRules.evaluate(RunRules.java:20)
    at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
    at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
    at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
    at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
    at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
    at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
    at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
    at org.junit.runners.Suite.runChild(Suite.java:128)
    at org.junit.runners.Suite.runChild(Suite.java:27)
    at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
    at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
    at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
    at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
    at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
    at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
    at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:264)
    at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:153)
    at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:124)
    at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:200)
    at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:153)
    at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:103)
Caused by: java.nio.file.FileSystemException: /opt/teamcity-agent/work/7c591132dbf26b78/enterprise/core-edge/target/test-data/org.neo4j.coreedge.scenarios.ConvertNonCoreEdgeStoreIT/F92F150B14412E95F08F8934BF39E639/server-edge-4/temp-copy/schema/label/lucene/labelStore/1/_1_BlockTreeOrds_750.tio: Too many open files
    at sun.nio.fs.UnixException.translateToIOException(UnixException.java:91)
    at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
    at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
    at sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:214)
    at java.nio.file.spi.FileSystemProvider.newOutputStream(FileSystemProvider.java:434)
    at java.nio.file.Files.newOutputStream(Files.java:216)
    at org.apache.lucene.store.FSDirectory$FSIndexOutput.<init>(FSDirectory.java:271)
    at org.apache.lucene.store.FSDirectory.createOutput(FSDirectory.java:224)
    at org.apache.lucene.store.NRTCachingDirectory.createOutput(NRTCachingDirectory.java:156)
    at org.apache.lucene.store.LockValidatingDirectoryWrapper.createOutput(LockValidatingDirectoryWrapper.java:44)
    at org.apache.lucene.store.TrackingDirectoryWrapper.createOutput(TrackingDirectoryWrapper.java:43)
    at org.apache.lucene.codecs.blocktreeords.OrdsBlockTreeTermsWriter.<init>(OrdsBlockTreeTermsWriter.java:188)
    at org.apache.lucene.codecs.blocktreeords.BlockTreeOrdsPostingsFormat.fieldsConsumer(BlockTreeOrdsPostingsFormat.java:75)
    at org.apache.lucene.codecs.perfield.PerFieldPostingsFormat$FieldsWriter.write(PerFieldPostingsFormat.java:196)
    at org.apache.lucene.index.FreqProxTermsWriter.flush(FreqProxTermsWriter.java:107)
    at org.apache.lucene.index.DefaultIndexingChain.flush(DefaultIndexingChain.java:126)
    at org.apache.lucene.index.DocumentsWriterPerThread.flush(DocumentsWriterPerThread.java:422)
    at org.apache.lucene.index.DocumentsWriter.doFlush(DocumentsWriter.java:503)
    at org.apache.lucene.index.DocumentsWriter.postUpdate(DocumentsWriter.java:377)
    at org.apache.lucene.index.DocumentsWriter.updateDocument(DocumentsWriter.java:467)
    at org.apache.lucene.index.IndexWriter.updateDocument(IndexWriter.java:1477)
    at org.neo4j.kernel.api.impl.labelscan.writer.PartitionedLuceneLabelScanWriter.flush(PartitionedLuceneLabelScanWriter.java:158)
    at org.neo4j.kernel.api.impl.labelscan.writer.PartitionedLuceneLabelScanWriter.write(PartitionedLuceneLabelScanWriter.java:90)
    at org.neo4j.kernel.impl.api.scan.LabelScanStoreProvider$FullLabelStream.visit(LabelScanStoreProvider.java:124)
    at org.neo4j.kernel.impl.api.scan.LabelScanStoreProvider$FullLabelStream.visit(LabelScanStoreProvider.java:99)
    at org.neo4j.kernel.impl.transaction.state.NeoStoreIndexStoreView$StoreViewNodeStoreScan.process(NeoStoreIndexStoreView.java:278)
    at org.neo4j.kernel.impl.transaction.state.NeoStoreIndexStoreView$NodeStoreScan.run(NeoStoreIndexStoreView.java:217)
    at org.neo4j.kernel.impl.api.scan.LabelScanStoreProvider$FullLabelStream.applyTo(LabelScanStoreProvider.java:117)
    at org.neo4j.kernel.api.impl.labelscan.LuceneLabelScanStore.rebuild(LuceneLabelScanStore.java:191)
    at org.neo4j.kernel.api.impl.labelscan.LuceneLabelScanStore.start(LuceneLabelScanStore.java:180)
    at org.neo4j.kernel.impl.storageengine.impl.recordstorage.RecordStorageEngine.start(RecordStorageEngine.java:435)
    at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:433)
    ... 62 more
    Suppressed: org.apache.lucene.store.AlreadyClosedException: this IndexWriter is closed
        at org.apache.lucene.index.IndexWriter.ensureOpen(IndexWriter.java:720)
        at org.apache.lucene.index.IndexWriter.ensureOpen(IndexWriter.java:734)
        at org.apache.lucene.index.IndexWriter.updateDocument(IndexWriter.java:1473)
        at org.neo4j.kernel.api.impl.labelscan.writer.PartitionedLuceneLabelScanWriter.flush(PartitionedLuceneLabelScanWriter.java:158)
        at org.neo4j.kernel.api.impl.labelscan.writer.PartitionedLuceneLabelScanWriter.close(PartitionedLuceneLabelScanWriter.java:100)
        at org.neo4j.kernel.api.impl.labelscan.LuceneLabelScanStore.rebuild(LuceneLabelScanStore.java:192)
        ... 65 more
    [CIRCULAR REFERENCE:java.nio.file.FileSystemException: /opt/teamcity-agent/work/7c591132dbf26b78/enterprise/core-edge/target/test-data/org.neo4j.coreedge.scenarios.ConvertNonCoreEdgeStoreIT/F92F150B14412E95F08F8934BF39E639/server-edge-4/temp-copy/schema/label/lucene/labelStore/1/_1_BlockTreeOrds_750.tio: Too many open files]
```

**Comment by mneedham** [comment]
@davidegrohmann  we were able to reproduce locally just running that test on a VM with low ulimit. I think on the build agents it's set to 4096 open files. We usually suggest 10x that for one server so I think we just need to bump it on the agents.

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/catchup/storecopy/edge/CopiedStoreRecovery.java [hunk 1]
```diff
@@ -46,8 +47,32 @@ public CopiedStoreRecovery( Config config, Iterable<KernelExtensionFactory<?>> k
...
-        GraphDatabaseService graphDatabaseService = newTempDatabase( tempStore );
-        graphDatabaseService.shutdown();
+        try
+        {
+            GraphDatabaseService graphDatabaseService = newTempDatabase( tempStore );
+            graphDatabaseService.shutdown();
+        }
+        catch ( Exception e )
+        {
+            if ( e.getCause().getCause() instanceof UpgradeNotAllowedByConfigurationException )
```
Comment (davidegrohmann) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/catchup/storecopy/edge/CopiedStoreRecovery.java:57]: shouldn't we check that the `cause` is not `null`?

Comment (mneedham) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/catchup/storecopy/edge/CopiedStoreRecovery.java:57]: @davidegrohmann  good idea. I'll add that as another commit. Want to merge this in now so @martinfurmanski  can do soak testing with the high limit record format

**Pull request merged**